File: src/main/services/mod-tools/touch-profile-types.ts
import type { BlendBoneInfo } from "@shared/body-shape";
import type { TouchProfileLlmSettings } from "@shared/touch-profile-llm";
import type { TouchZoneSettings } from "@shared/touch-profile-settings";

export const TOUCH_RUNTIME_VERSION = "1";
export const TOUCH_PROMPT_VERSION = "17";
export const TOUCH_VISION_CACHE_VERSION = "15";
export const TOUCH_VISION_EVALUATOR_VERSION = "1";
export const TOUCH_VISION_MASK_TUNING_VERSION = "2";
export const TOUCH_VISION_SEED_INFLUENCE_SCALE_RANGE = {
    min: 0.65,
    max: 1.25,
    default: 1,
    step: 0.05,
} as const;
export const TOUCH_VISION_MASK_TUNING_RANGES = {
    radiusPadding: { min: 0.95, max: 1.12, step: 0.01 },
    maskCutoffD2: { min: 1, max: 3.5, step: 0.05 },
    maskEdgeFadeD2: { min: 0.05, max: 1.25, step: 0.05 },
    colocatedLayerRatio: { min: 0, max: 0.5, step: 0.01 },
    colocatedLayerMin: { min: 0, max: 0.05, step: 0.001 },
    sideHeightPadRatio: { min: 0, max: 0.25, step: 0.01 },
    sideHeightPadMin: { min: 0, max: 0.08, step: 0.001 },
    seedRadiusScale: { min: 0.9, max: 1.15, step: 0.01 },
} as const;
export const DEFAULT_TOUCH_VISION_MASK_TUNING = {
    radiusPadding: 1.02,
    maskCutoffD2: 2.25,
    maskEdgeFadeD2: 0.55,
    seedInfluenceScale: TOUCH_VISION_SEED_INFLUENCE_SCALE_RANGE.default,
    colocatedLayerRatio: 0.28,
    colocatedLayerMin: 0.012,
    sideHeightPadRatio: 0.12,
    sideHeightPadMin: 0.02,
    seedRadiusScale: 1.05,
} as const;
/** Maximum number of position Vision pipelines running at once. */
export const TOUCH_VISION_CONCURRENCY = 3;
export const TOUCH_PROFILE_MANIFEST_FILE = ".nahida-touch-profile.json";
export const TOUCH_PROFILE_MANIFEST_KIND = "nahida-touch-profile";
export const TOUCH_OBJECT_MODE = 7;
export const TOUCH_POSITION_STRIDE = 40;
export const TOUCH_MASK_BANDS = 3;
export const TOUCH_ZONE_CHANNELS = 12;
export const TOUCH_BAKE_SAMPLES = 8;
export const TOUCH_FOLDER_SUFFIX = " (Touch)";
/** Minimum confidence on every interactive component. */
export const TOUCH_CONFIDENCE_AUTO_APPLY_MIN = 0.55;
/** Average confidence across interactive components. */
export const TOUCH_CONFIDENCE_AUTO_APPLY_AVG = 0.65;

export const TOUCH_SHADER_FILES = [
    "rzm_gs_probe.hlsl",
    "rzm_object_detect.hlsl",
    "rzm_pin_detected.hlsl",
    "rzm_jiggle_screen_state.hlsl",
    "rzm_jiggle_interaction.hlsl",
] as const;

export type TouchSupportGrade = "A" | "B" | "C";
export type TouchComponentKind = "body" | "legs" | "hair" | "accessory" | "unknown";
export type TouchProgressStage =
    | "scan"
    | "preview"
    | "vision"
    | "assets"
    | "ini"
    | "validate"
    | "complete";

export type TouchDrawRange = {
    firstIndex: number;
    indexCount: number;
    baseVertex: number;
    label?: string;
    conditionText?: string;
};

export type TouchObjectMapEntry = {
    firstIndex: number;
    indexCount: number;
    objectMode: number;
    objectId: number;
    label: string;
};

export type TouchComponentAnalysis = {
    id: string;
    name: string;
    kind: TouchComponentKind;
    interactiveCandidate: boolean;
    supportGrade: TouchSupportGrade;
    supportReasons: string[];
    positionResourceName: string;
    positionRelativePath: string;
    positionPath: string;
    positionStride: number;
    vertexCount: number;
    indexResourceName?: string;
    indexRelativePath?: string;
    indexPath?: string;
    indexRelativePaths?: string[];
    indexPaths?: string[];
    indexFormat?: string;
    indexCount: number;
    blendSectionName?: string;
    ibSectionName?: string;
    ibHash?: string;
    variantKey?: string;
    variantCondition?: string;
    drawRanges: TouchDrawRange[];
    objectMaps: TouchObjectMapEntry[];
    blendRelativePath?: string;
    blendPath?: string;
    blendStride?: number;
    bones: BlendBoneInfo[];
};

export type TouchModAnalysis = {
    /** Directory containing the resolved INI (may be a body/face subfolder). */
    modRoot: string;
    /** Directory the user selected; used for output naming and source disabling. */
    sourceRoot: string;
    /** modRoot relative to sourceRoot; reapplies to the renamed source after disable. */
    modRootRelativeToSource: string;
    iniPath: string;
    iniRelativePath: string;
    sourceFilesRelativePaths: string[];
    supportGrade: TouchSupportGrade;
    supportReasons: string[];
    components: TouchComponentAnalysis[];
    meshHash: string;
    iniHash: string;
};

export type TouchComponentInspection = {
    id: string;
    name: string;
    kind: TouchComponentKind;
    supportGrade: TouchSupportGrade;
    interactiveCandidate: boolean;
    vertexCount: number;
    indexCount: number;
    variantKey?: string;
    variantCondition?: string;
    objectMaps: TouchObjectMapEntry[];
    hasBlend: boolean;
    bones: BlendBoneInfo[];
};

export type TouchModInspection = {
    sessionId: string;
    modRoot: string;
    iniRelativePath: string;
    sourceFilesRelativePaths: string[];
    supportGrade: TouchSupportGrade;
    supportReasons: string[];
    components: TouchComponentInspection[];
};

export type TouchVisionZone = {
    id: string;
    label: string;
    confidence: number;
    maskTuning?: Partial<TouchVisionMaskTuning>;
    include: Record<string, Array<Array<[number, number]>>>;
    exclude: Record<string, Array<Array<[number, number]>>>;
};

export type TouchVisionMaskTuning = {
    radiusPadding: number;
    maskCutoffD2: number;
    maskEdgeFadeD2: number;
    seedInfluenceScale: number;
    colocatedLayerRatio: number;
    colocatedLayerMin: number;
    sideHeightPadRatio: number;
    sideHeightPadMin: number;
    seedRadiusScale: number;
};

export type TouchVisionResult = {
    componentId: string;
    isHumanBody: boolean;
    approved: boolean;
    interactive: boolean;
    zones: TouchVisionZone[];
    excludedRegions: string[];
    warnings: string[];
};

export type TouchZoneSpec = {
    id: string;
    label: string;
    channel: number;
    confidence: number;
    center: [number, number, number];
    radius: [number, number, number];
    source: "vision" | "manual" | "bone";
    settings: TouchZoneSettings;
    /**
     * Vertices selected from vision polygons (any view). When present, soft masks are
     * built from distance-to-seed rather than a pure geometric ellipsoid.
     */
    seedVertices?: number[];
    /** Bounded Vision-only mask tuning; manual zones leave this unset. */
    visionMaskTuning?: Partial<TouchVisionMaskTuning>;
};

export type TouchTurnRecord = {
    turn: number;
    vision: TouchVisionResult;
    zones: TouchZoneSpec[];
    confidence: number;
    warnings: string[];
    approved: boolean;
};

export type TouchComponentDraft = {
    componentId: string;
    interactive: boolean;
    objectId: number;
    zones: TouchZoneSpec[];
    vision?: TouchVisionResult;
    visionApproved?: boolean;
    previewImageRelativePath?: string;
    confidence: number;
    warnings: string[];
    currentTurn?: number;
    turnHistory?: TouchTurnRecord[];
};

export type TouchDraft = {
    sessionId: string;
    createdAt: string;
    sourceModRoot: string;
    analysis: TouchModAnalysis;
    components: TouchComponentDraft[];
    visionUsed: boolean;
    modelName: string;
    llm: TouchProfileLlmSettings;
    promptVersion: string;
    runtimeVersion: string;
    canAutoApply: boolean;
    warnings: string[];
};

export type TouchProgressEvent = {
    sessionId?: string;
    stage: TouchProgressStage;
    progress: number;
    message: string;
    componentId?: string;
};

export type TouchApplyResult = {
    sessionId: string;
    outputModRoot: string;
    /** Source mod path after apply (disabled when reenableSourceOnRollback is true). */
    sourceModRoot: string;
    /** True when apply disabled the source; rollback should re-enable it. */
    reenableSourceOnRollback: boolean;
    disabled: boolean;
    validation: TouchValidationResult;
    warnings: string[];
};

export type TouchRollbackResult = {
    outputModRoot: string;
    sourceModRoot: string;
    removedOutput: boolean;
    reenabledSource: boolean;
};

export type TouchValidationIssue = {
    level: "error" | "warning";
    code: string;
    message: string;
    componentId?: string;
};

export type TouchValidationResult = {
    ok: boolean;
    issues: TouchValidationIssue[];
};

export type TouchJiggleParams = {
    objectId: number;
    radius: number;
    strength: number;
    falloff: number;
    dragScale: number;
    grabDamping: number;
    grabSpring: number;
    releaseDamping: number;
    releaseSpring: number;
    releaseKick: number;
    maxOffset: number;
    targetFollow: number;
    mouseYDirection: number;
    mouseXDirection: number;
};

export const DEFAULT_TOUCH_JIGGLE_PARAMS: Omit<TouchJiggleParams, "objectId"> = {
    radius: 0.2,
    strength: 1.15,
    falloff: 1.8,
    dragScale: 1,
    grabDamping: 0.86,
    grabSpring: 0.176,
    releaseDamping: 0.96,
    releaseSpring: 0.055,
    releaseKick: 1.18,
    maxOffset: 0.065,
    targetFollow: 0.12,
    mouseYDirection: 1,
    mouseXDirection: 1,
};

File: internal/tools/touch_profile_types.go
package tools

const (
	touchRuntimeVersion        = "1"
	touchPromptVersion         = "17"
	touchProfileManifestFile   = ".nahida-touch-profile.json"
	touchProfileManifestKind   = "nahida-touch-profile"
	touchObjectMode            = 7
	touchPositionStride        = 40
	touchMaskBands             = 3
	touchZoneChannels          = 12
	touchBakeSamples           = 8
	touchFolderSuffix          = " (Touch)"
	touchConfidenceAutoMin     = 0.55
	touchConfidenceAutoAverage = 0.65
)

var touchShaderFiles = []string{
	"rzm_gs_probe.hlsl",
	"rzm_object_detect.hlsl",
	"rzm_pin_detected.hlsl",
	"rzm_jiggle_screen_state.hlsl",
	"rzm_jiggle_interaction.hlsl",
}

type TouchDrawRange struct {
	FirstIndex    int     `json:"firstIndex"`
	IndexCount    int     `json:"indexCount"`
	BaseVertex    int     `json:"baseVertex"`
	Label         *string `json:"label,omitempty"`
	ConditionText *string `json:"conditionText,omitempty"`
}

type TouchObjectMapEntry struct {
	FirstIndex int    `json:"firstIndex"`
	IndexCount int    `json:"indexCount"`
	ObjectMode int    `json:"objectMode"`
	ObjectID   int    `json:"objectId"`
	Label      string `json:"label"`
}

type TouchComponentAnalysis struct {
	ID                   string                `json:"id"`
	Name                 string                `json:"name"`
	Kind                 string                `json:"kind"`
	InteractiveCandidate bool                  `json:"interactiveCandidate"`
	SupportGrade         string                `json:"supportGrade"`
	SupportReasons       []string              `json:"supportReasons"`
	PositionResourceName string                `json:"positionResourceName"`
	PositionRelativePath string                `json:"positionRelativePath"`
	PositionPath         string                `json:"positionPath"`
	PositionStride       int                   `json:"positionStride"`
	VertexCount          int                   `json:"vertexCount"`
	IndexResourceName    *string               `json:"indexResourceName,omitempty"`
	IndexRelativePath    *string               `json:"indexRelativePath,omitempty"`
	IndexPath            *string               `json:"indexPath,omitempty"`
	IndexRelativePaths   []string              `json:"indexRelativePaths,omitempty"`
	IndexPaths           []string              `json:"indexPaths,omitempty"`
	IndexFormats         []*string             `json:"indexFormats,omitempty"`
	IndexFormat          *string               `json:"indexFormat,omitempty"`
	IndexCount           int                   `json:"indexCount"`
	BlendSectionName     *string               `json:"blendSectionName,omitempty"`
	IBSectionName        *string               `json:"ibSectionName,omitempty"`
	IBHash               *string               `json:"ibHash,omitempty"`
	VariantKey           *string               `json:"variantKey,omitempty"`
	VariantCondition     *string               `json:"variantCondition,omitempty"`
	DrawRanges           []TouchDrawRange      `json:"drawRanges"`
	ObjectMaps           []TouchObjectMapEntry `json:"objectMaps"`
	BlendRelativePath    *string               `json:"blendRelativePath,omitempty"`
	BlendPath            *string               `json:"blendPath,omitempty"`
	BlendStride          *int                  `json:"blendStride,omitempty"`
	Bones                []BlendBoneInfo       `json:"bones"`
}

type TouchModAnalysis struct {
	ModRoot                  string                   `json:"modRoot"`
	SourceRoot               string                   `json:"sourceRoot"`
	ModRootRelativeToSource  string                   `json:"modRootRelativeToSource"`
	INIPath                  string                   `json:"iniPath"`
	INIRelativePath          string                   `json:"iniRelativePath"`
	SourceFilesRelativePaths []string                 `json:"sourceFilesRelativePaths"`
	SupportGrade             string                   `json:"supportGrade"`
	SupportReasons           []string                 `json:"supportReasons"`
	Components               []TouchComponentAnalysis `json:"components"`
	MeshHash                 string                   `json:"meshHash"`
	INIHash                  string                   `json:"iniHash"`
}

type TouchComponentInspection struct {
	ID                   string                `json:"id"`
	Name                 string                `json:"name"`
	Kind                 string                `json:"kind"`
	SupportGrade         string                `json:"supportGrade"`
	InteractiveCandidate bool                  `json:"interactiveCandidate"`
	VertexCount          int                   `json:"vertexCount"`
	IndexCount           int                   `json:"indexCount"`
	VariantKey           *string               `json:"variantKey,omitempty"`
	VariantCondition     *string               `json:"variantCondition,omitempty"`
	ObjectMaps           []TouchObjectMapEntry `json:"objectMaps"`
	HasBlend             bool                  `json:"hasBlend"`
	Bones                []BlendBoneInfo       `json:"bones"`
}

type TouchModInspection struct {
	SessionID                string                     `json:"sessionId"`
	ModRoot                  string                     `json:"modRoot"`
	INIRelativePath          string                     `json:"iniRelativePath"`
	SourceFilesRelativePaths []string                   `json:"sourceFilesRelativePaths"`
	SupportGrade             string                     `json:"supportGrade"`
	SupportReasons           []string                   `json:"supportReasons"`
	Components               []TouchComponentInspection `json:"components"`
}

type TouchAdvancedSettings struct {
	Radius    float64 `json:"radius"`
	Strength  float64 `json:"strength"`
	Damping   float64 `json:"damping"`
	Spring    float64 `json:"spring"`
	MaxOffset float64 `json:"maxOffset"`
	Falloff   float64 `json:"falloff"`
}

type TouchZoneSettings struct {
	MaskStrength        float64               `json:"maskStrength"`
	MaskCurve           float64               `json:"maskCurve"`
	MaskRadiusScale     float64               `json:"maskRadiusScale"`
	MaskCoreAttenuation string                `json:"maskCoreAttenuation"`
	StrengthPreset      string                `json:"strengthPreset"`
	PhysicsPreset       string                `json:"physicsPreset"`
	Advanced            TouchAdvancedSettings `json:"advanced"`
}

type TouchZoneSpec struct {
	ID         string            `json:"id"`
	Label      string            `json:"label"`
	Channel    int               `json:"channel"`
	Confidence float64           `json:"confidence"`
	Center     [3]float64        `json:"center"`
	Radius     [3]float64        `json:"radius"`
	Source     string            `json:"source"`
	Settings   TouchZoneSettings `json:"settings"`
	Seeds      []int             `json:"seedVertices,omitempty"`
}

type TouchComponentDraft struct {
	ComponentID string          `json:"componentId"`
	Interactive bool            `json:"interactive"`
	ObjectID    int             `json:"objectId"`
	Zones       []TouchZoneSpec `json:"zones"`
	Confidence  float64         `json:"confidence"`
	Warnings    []string        `json:"warnings"`
}

type TouchProfileLLMSettings struct {
	Protocol  string `json:"protocol"`
	Endpoint  string `json:"endpoint"`
	Model     string `json:"model"`
	Reasoning string `json:"reasoning"`
}

type TouchDraft struct {
	SessionID      string                  `json:"sessionId"`
	CreatedAt      string                  `json:"createdAt"`
	SourceModRoot  string                  `json:"sourceModRoot"`
	Analysis       TouchModAnalysis        `json:"analysis"`
	Components     []TouchComponentDraft   `json:"components"`
	VisionUsed     bool                    `json:"visionUsed"`
	ModelName      string                  `json:"modelName"`
	LLM            TouchProfileLLMSettings `json:"llm"`
	PromptVersion  string                  `json:"promptVersion"`
	RuntimeVersion string                  `json:"runtimeVersion"`
	CanAutoApply   bool                    `json:"canAutoApply"`
	Warnings       []string                `json:"warnings"`
}

type TouchProgressEvent struct {
	SessionID   *string `json:"sessionId,omitempty"`
	Stage       string  `json:"stage"`
	Progress    float64 `json:"progress"`
	Message     string  `json:"message"`
	ComponentID *string `json:"componentId,omitempty"`
}

type TouchMeshDescriptor struct {
	SessionID        string          `json:"sessionId"`
	ComponentID      string          `json:"componentId"`
	TopologyRevision string          `json:"topologyRevision"`
	VertexCount      int             `json:"vertexCount"`
	PositionsURL     string          `json:"positionsUrl"`
	PositionsCount   int             `json:"positionsCount"`
	IndicesURL       *string         `json:"indicesUrl,omitempty"`
	IndexCount       int             `json:"indexCount"`
	Bones            []BlendBoneInfo `json:"bones"`
	BlendStride      *int            `json:"blendStride,omitempty"`
	BlendURL         *string         `json:"blendUrl,omitempty"`
	BlendBytes       int             `json:"blendBytes"`
}

type TouchPreviewZoneDescriptor struct {
	TouchZoneSpec
	WeightOffset int `json:"weightOffset"`
}

type TouchProfilePreviewDescriptor struct {
	SessionID       string                       `json:"sessionId"`
	ComponentID     string                       `json:"componentId"`
	PreviewRevision uint64                       `json:"previewRevision"`
	VertexCount     int                          `json:"vertexCount"`
	WeightsURL      string                       `json:"weightsUrl"`
	WeightsCount    int                          `json:"weightsCount"`
	Zones           []TouchPreviewZoneDescriptor `json:"zones"`
}

type TouchValidationIssue struct {
	Level       string  `json:"level"`
	Code        string  `json:"code"`
	Message     string  `json:"message"`
	ComponentID *string `json:"componentId,omitempty"`
}

type TouchValidationResult struct {
	OK     bool                   `json:"ok"`
	Issues []TouchValidationIssue `json:"issues"`
}

type TouchApplyResult struct {
	SessionID                string                `json:"sessionId"`
	OutputModRoot            string                `json:"outputModRoot"`
	SourceModRoot            string                `json:"sourceModRoot"`
	ReenableSourceOnRollback bool                  `json:"reenableSourceOnRollback"`
	Disabled                 bool                  `json:"disabled"`
	Validation               TouchValidationResult `json:"validation"`
	Warnings                 []string              `json:"warnings"`
}

type TouchRollbackResult struct {
	OutputModRoot   string `json:"outputModRoot"`
	SourceModRoot   string `json:"sourceModRoot"`
	RemovedOutput   bool   `json:"removedOutput"`
	ReenabledSource bool   `json:"reenabledSource"`
}

type TouchJiggleParams struct {
	ObjectID        int     `json:"objectId"`
	Radius          float64 `json:"radius"`
	Strength        float64 `json:"strength"`
	Falloff         float64 `json:"falloff"`
	DragScale       float64 `json:"dragScale"`
	GrabDamping     float64 `json:"grabDamping"`
	GrabSpring      float64 `json:"grabSpring"`
	ReleaseDamping  float64 `json:"releaseDamping"`
	ReleaseSpring   float64 `json:"releaseSpring"`
	ReleaseKick     float64 `json:"releaseKick"`
	MaxOffset       float64 `json:"maxOffset"`
	TargetFollow    float64 `json:"targetFollow"`
	MouseYDirection float64 `json:"mouseYDirection"`
	MouseXDirection float64 `json:"mouseXDirection"`
}

var defaultTouchJiggleParams = TouchJiggleParams{
	Radius: 0.2, Strength: 1.15, Falloff: 1.8, DragScale: 1,
	GrabDamping: 0.86, GrabSpring: 0.176, ReleaseDamping: 0.96,
	ReleaseSpring: 0.055, ReleaseKick: 1.18, MaxOffset: 0.065,
	TargetFollow: 0.12, MouseYDirection: 1, MouseXDirection: 1,
}
